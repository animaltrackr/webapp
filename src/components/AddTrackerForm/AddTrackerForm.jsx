import React, { Component } from 'react';

import { Divider, Typography, Button, Icon, Spin, Input } from 'antd';

import * as api from 'modules/api';

import Options from '../Options/Options';
import MaxErrorInput from '../MaxErrorInput/MaxErrorInput';

import './AddTrackerForm.less';

const { Text, Title } = Typography;

class AddTrackerForm extends Component {
	constructor(props) {
		super(props);

		this.state = {
			name: '',
			status: this.props.option ? this.props.options.statusOptions[0][0] : 'A',
			colour: this.props.option
				? this.props.options.colourOptions[0][0]
				: 'red',
			max_error_radius: 0,
			location_method: this.props.options
				? this.props.options.locationMethodOptions[0][0]
				: 'G',
			isLoading: false,
		};
	}

	onNameChange = e => {
		this.setState({ name: e.target.value });
	};

	onStateChange = (value, type) => {
		if (type === 'colour') {
			this.setState({ colour: value });
		} else if (type === 'status') {
			this.setState({ status: value });
		} else if (type === 'location_method') {
			this.setState({ location_method: value });
		}
	};

	createNewDeer = id => {
		const newDeer = {
			name: this.state.name,
			id: id,
			status: this.state.status,
			visible: false,
			colour: this.state.colour,
			points: [],
			max_error_radius: this.state.max_error_radius,
			location_method: this.state.location_method,
			loading: {
				status: false,
				name: false,
				colour: false,
				location_method: false,
				max_error_radius: false,
			},
		};
		this.props.addDeerToDeerStates(newDeer);
	};

	handleSubmit = () => {
		this.setState({ isLoading: true });

		api
			.createTracker({
				animal_id: this.state.name,
				status: this.state.status,
				max_error_radius: this.state.max_error_radius,
				location_method: this.state.location_method,
			})
			.then(values => {
				this.createNewDeer(values.id);
				this.handleClear();
				this.props.toggleDrawer();
			});
	};

	handleClear = () => {
		this.setState(
			{
				name: '',
				status: this.props.options.statusOptions[0][0],
				colour: this.props.options.colourOptions[0][0],
				max_error_radius: 0,
				location_method: this.props.options.locationMethodOptions[0][0],
			},
			() => this.setState({ isLoading: false })
		);
	};

	onMaxErrorRadiusChange = value => {
		this.setState({ max_error_radius: value });
	};

	render() {
		const loadingIcon = <Icon type="loading" style={{ fontSize: 20 }} spin />;

		return (
			<div className="tracker-form">
				<Title level={4}>Add New Tracker</Title>
				<Divider />
				<div className="input">
					<div className="input-title-wrapper">
						<Text strong>Name:</Text>
						<div className="input-wrapper">
							<Input
								allowClear
								onChange={this.onNameChange}
								placeholder="Enter Name"
							/>
						</div>
					</div>
					<Options
						onStateChange={this.onStateChange}
						title="Status"
						options={this.props.options ? this.props.options.statusOptions : ''}
						default={this.state.status}
						type="status"
					/>
					<Options
						onStateChange={this.onStateChange}
						title="Colour"
						options={this.props.options ? this.props.options.colourOptions : ''}
						default={this.state.colour}
						type="colour"
					/>
					<Options
						onStateChange={this.onStateChange}
						title="Location Method"
						options={
							this.props.options ? this.props.options.locationMethodOptions : ''
						}
						default={this.state.location_method}
						type="location_method"
					/>
					<Text>Maximum Error Radius:</Text>
					<MaxErrorInput
						onMaxErrorRadiusChange={this.onMaxErrorRadiusChange}
						maxErrorRadius={this.state.max_error_radius}
					/>
				</div>
				<div className="button-wrapper">
					<div className="submit-button">
						<Button type="primary" onClick={this.handleSubmit}>
							Submit
						</Button>
						<div className="spin-icon-wrapper">
							<Spin indicator={loadingIcon} hidden={!this.state.isLoading} />
						</div>
					</div>
				</div>
			</div>
		);
	}
}

AddTrackerForm.propTypes = {};

export default AddTrackerForm;
