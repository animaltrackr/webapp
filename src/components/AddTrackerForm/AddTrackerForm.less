.add-form .input {
    padding: 0;
    margin: 0;
    display: flex;
    align-items: center;

}

.input-title-wrapper .input-wrapper {
    padding-left: 50px;
}

.tracker-form .button-wrapper {
    display: flex;
    padding-top: 25px;
}

.input-title-wrapper {
    margin: 0;
    display: flex;
    align-items: center;
    padding-bottom: 11px;
}

.submit-button {
    padding-right: 10px;
    margin: 0;
    display: flex;
    align-items: center;
}

.spin-icon-wrapper {
    padding-left: 10px;
}