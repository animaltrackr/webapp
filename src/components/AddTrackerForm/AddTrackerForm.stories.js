import React from 'react';
import { storiesOf } from '@storybook/react';
import AddTrackerForm from './AddTrackerForm';

storiesOf('AddTrackerForm', module).add('AddTrackerForm', () => (
	<div>
		<AddTrackerForm />
	</div>
));
