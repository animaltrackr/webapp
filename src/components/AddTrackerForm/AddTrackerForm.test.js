import React from 'react';
import { shallow } from 'enzyme';
import AddTrackerForm from './AddTrackerForm';

describe('<AddTrackerForm />', () => {
	test('renders', () => {
		const wrapper = shallow(<AddTrackerForm />);
		expect(wrapper).toMatchSnapshot();
	});
});
