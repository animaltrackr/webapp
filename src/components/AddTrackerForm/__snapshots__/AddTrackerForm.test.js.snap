// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTrackerForm /> renders 1`] = `ShallowWrapper {}`;
